### PR TITLE
Fix permalink generation and handling on non En

### PR DESCRIPTION
**Pull request by lucatume**
Issues: 

* https://moderntribe.atlassian.net/browse/TEC-3479
* https://moderntribe.atlassian.net/browse/ECP-427
* https://moderntribe.atlassian.net/browse/ECP-422

Based on: https://github.com/moderntribe/tribe-common/pull/1402

This PR fixes how we generate and consume translation informations for
non English websites when dealing with rewrite rules and, hence, with
pretty permalinks and URLs.
This PR does affect the way PRO code works by changing some basic functions
PRO relies on; in the screencast you can see I'm testing PRO views too and
they work.

There are 4 key changes in this PR:

1. When generating the bases for our rewrite rules, take translations into
account; what would happen before was that, as an example some permalink
parts would not be translated (e.g. `/events`) leading to weird and non-working
permalinks like `/events/ヶ月` (Month view in a Japanese site).
2.The same work could be translated differently in diff. plugins, e.g. "month" is
translated with `ヶ月` in TEC and with `月` in PRO. This would lead to, again,
permalink breakages that this PR handles by including all translations in our rewrite
rules.
3. The link to the Month view on a Japanese site would be `/events/月/`, which presents
an half-translated URL; while this **keeps working**, the PR makes it so the Views
will generate, and the View backend will consume, fully-localized URLs; e.g. 
`/イベント/ヶ月/` is the same as `/events/ヶ月/` which is the same as `/events/month`:
a link to Month View.
4. Finally the PR introduces a bit more flexibility in the compilation of translations
to allow for flags, see the code for changes. The change is backward compatible and is
immediately leveraged in our code to ensure we compile the correct translations.

[Screencap](https://drive.google.com/open?id=19dhL3vLC9blUNaRamlH1VrhzWOVHc-M3&authuser=luca%40tri.be&usp=drive_fs)

### CODE REVIEW

File: src/Tribe/I18n.php [hunk 1]
```diff
@@ -45,15 +73,25 @@ public function __construct( TEC $tec ) {
...
 	 *
-	 * @param array  $strings          An array of strings (required).
-	 * @param array  $languages        Which l10n to fetch the string (required).
-	 * @param array  $domains          Possible Domains to re-load.
-	 * @param string $default_language The default language to avoid re-doing that.
+	 * @param array  $strings                     An array of strings (required).
+	 * @param array  $languages                   Which l10n to fetch the string (required).
+	 * @param array  $domains                     Possible Domains to re-load.
+	 * @param string $default_language            The default language to avoid re-doing that.
+	 * @param int $flags                          An integer resulting from the combination of compilation flags;
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:82]: :no_entry_sign: Expected 4 spaces after parameter type; 1 found

<sub>hash: cbc53eed3e1874c5edbc6cff1c79bf93</sub>

File: src/Tribe/I18n.php [hunk 2]
```diff
@@ -73,16 +111,25 @@ public function get_i18n_strings( $strings, $languages, $domains = array(), $def
 	 * Differently from the `get_i18n_strings` method this will not use any domain that's not specified.
 	 *
 	 * @since 5.1.1
+	 * @since TBD   Add support for the $flags argument.
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:114]: :no_entry_sign: There must be a single blank line after a tag group

<sub>hash: d9c3b09cbadb0540d8a300f4bcbd41cb</sub>

File: src/Tribe/I18n.php [hunk 3]
```diff
@@ -150,14 +197,30 @@ protected function with_locale( $locale, callable $do, array $args = [] ) {
...
+	 * translations of the strings in their `strtolower` versions.
+	 * Combine the flags using the usual PHP syntax: `I18n::COMPILE_INPUT | I18n::COMPILE_STRTOLOWER` to compile
+	 * only the translation of the string as input and in their lowercase version.
+	 *
 	 * @since 5.1.1
+	 * @since TBD   Add support for the $flags argument.
 	 *
 	 * @param array<string,array|string> $strings The set of strings to compile the translations for.
 	 * @param string|array<string>       $domains The domain(s) that should be used to compile the string translations.
+	 * @param int $flags                          An integer resulting from the combination of compilation flags;
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:211]: :no_entry_sign: Expected 24 spaces after parameter type; 1 found

<sub>hash: 236669cbda91f19fcd8308cc7eb7ff7e</sub>

File: src/Tribe/I18n.php [hunk 4]
```diff
@@ -150,14 +197,30 @@ protected function with_locale( $locale, callable $do, array $args = [] ) {
...
+	 *                                            as input.
+	 *                                            `static::COMPILE_STRTOLOWER` will compile the translation for the
+	 *                                            string in its lowercase version.
+	 *                                            `static::COMPILE_UCFIRST` will compile the translation for the string
+	 *                                            in its title version.
 	 *
 	 * @return array<string|array> A map of the compiled string translations.
 	 */
-	 public function compile_translations( array $strings, $domains ) {
+	 public function compile_translations( array $strings, $domains, $flags = 7 ) {
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:223]: * :warning: Found precision alignment of 1 spaces.
* :no_entry_sign: Line indented incorrectly; expected 1 tabs, found 1

<sub>hash: 1a1ec1b2dd333d2603fb82c4794f9617</sub>

File: src/Tribe/I18n.php [hunk 5]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
 
 				// Grab the possible strings for default and any other domain.
 				if ( 'default' === $domain ) {
-					$strings[ $key ][] = __( $value );
-					$strings[ $key ][] = __( strtolower( $value ) );
-					$strings[ $key ][] = __( ucfirst( $value ) );
+					if ($flags & static::COMPILE_INPUT) {
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:256]: * :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: No space before closing parenthesis is prohibited

<sub>hash: e3272186b0e6417ce2857329d343fde0</sub>

File: src/Tribe/I18n.php [hunk 6]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+						$strings[ $key ][] = __( $value );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:257]: :no_entry_sign: The $text arg must be a single string literal, not "$value".

<sub>hash: 0c7f8aa2dce2a7a85a549c5077b803a2</sub>

File: src/Tribe/I18n.php [hunk 7]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ($flags & static::COMPILE_STRTOLOWER) {
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:259]: * :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: No space before closing parenthesis is prohibited

<sub>hash: 18850de1916b692329436e561a269031</sub>

File: src/Tribe/I18n.php [hunk 8]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+						$strings[ $key ][] = __( strtolower( $value ) );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:260]: :no_entry_sign: The $text arg must be a single string literal, not "strtolower( $value )".

<sub>hash: cebf7d378d4a94ea0f9f4a9b51e2e3db</sub>

File: src/Tribe/I18n.php [hunk 9]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ($flags & static::COMPILE_UCFIRST) {
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:262]: * :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: No space before closing parenthesis is prohibited

<sub>hash: 8ded51bffe35ed70e5ba765be03f7a6f</sub>

File: src/Tribe/I18n.php [hunk 10]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+						$strings[ $key ][] = __( ucfirst( $value ) );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:263]: :no_entry_sign: The $text arg must be a single string literal, not "ucfirst( $value )".

<sub>hash: d4f84976fac2e4043c4cc881147a4e25</sub>

File: src/Tribe/I18n.php [hunk 11]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
 				} else {
-					$strings[ $key ][] = __( $value, $domain );
-					$strings[ $key ][] = __( strtolower( $value ), $domain );
-					$strings[ $key ][] = __( ucfirst( $value ), $domain );
+					if ($flags & static::COMPILE_INPUT) {
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:266]: * :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: No space before closing parenthesis is prohibited

<sub>hash: ecd55bf162e2335cf8a4211b376e2795</sub>

File: src/Tribe/I18n.php [hunk 12]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+						$strings[ $key ][] = __( $value, $domain );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:267]: * :no_entry_sign: The $text arg must be a single string literal, not "$value".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: 47d465761e59147883a12f6b16fe1374</sub>

File: src/Tribe/I18n.php [hunk 13]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ($flags & static::COMPILE_STRTOLOWER) {
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:269]: * :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: No space before closing parenthesis is prohibited

<sub>hash: 43b91b4f56c0864ade42dc60961d2072</sub>

File: src/Tribe/I18n.php [hunk 14]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+						$strings[ $key ][] = __( strtolower( $value ), $domain );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:270]: * :no_entry_sign: The $text arg must be a single string literal, not "strtolower( $value )".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: 3a43ef1ee287f718aa5a943c4712e875</sub>

File: src/Tribe/I18n.php [hunk 15]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ($flags & static::COMPILE_UCFIRST) {
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:272]: * :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: No space before closing parenthesis is prohibited

<sub>hash: 5e967a59336058c96b6ea5de4817f11e</sub>

File: src/Tribe/I18n.php [hunk 16]
```diff
@@ -190,13 +253,25 @@ public function compile_translations( array $strings, $domains ) {
...
+						$strings[ $key ][] = __( ucfirst( $value ), $domain );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:273]: * :no_entry_sign: The $text arg must be a single string literal, not "ucfirst( $value )".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: cbf0a522be795d72ccfdbc921527a970</sub>

File: src/Tribe/Rewrite.php [hunk 17]
```diff
@@ -731,7 +735,9 @@ public function get_localized_bases( array $bases, array $domains ) {
 			return $cached;
 		}
 
-		$localized_bases = tribe( 'tec.i18n' )->get_i18n_strings_for_domains( $bases, [ $locale ], $domains );
+		$flags = I18n::COMPILE_STRTOLOWER;
```
Comment (tr1b0t) [inline src/Tribe/Rewrite.php:738]: :warning: Equals sign not aligned with surrounding assignments; expected 11 spaces but found 1 space

<sub>hash: 154b39f3e54764543e7914bb99b54d74</sub>

File: src/Tribe/I18n.php [hunk 18]
```diff
@@ -150,14 +194,30 @@ protected function with_locale( $locale, callable $do, array $args = [] ) {
...
+	 * translations of the strings in their `strtolower` versions.
+	 * Combine the flags using the usual PHP syntax: `I18n::COMPILE_INPUT | I18n::COMPILE_STRTOLOWER` to compile
+	 * only the translation of the string as input and in their lowercase version.
+	 *
 	 * @since 5.1.1
+	 * @since TBD   Add support for the $flags argument.
 	 *
 	 * @param array<string,array|string> $strings The set of strings to compile the translations for.
 	 * @param string|array<string>       $domains The domain(s) that should be used to compile the string translations.
+     * @param int                        $flags   An integer resulting from the combination of compilation flags;
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:208]: :no_entry_sign: Tabs must be used to indent lines; spaces are not allowed

<sub>hash: 00cdfc0e9fc833ec79192276e1e41a6b</sub>

File: src/Tribe/I18n.php [hunk 19]
```diff
@@ -150,14 +194,30 @@ protected function with_locale( $locale, callable $do, array $args = [] ) {
...
+     *                                            defaults to `static::COMPILE_ALL` to compile all versions of the
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:209]: :no_entry_sign: Tabs must be used to indent lines; spaces are not allowed

<sub>hash: fdcf25a4f21c2e0ad85bf7cc28a5ee8a</sub>

File: src/Tribe/I18n.php [hunk 20]
```diff
@@ -190,13 +250,25 @@ public function compile_translations( array $strings, $domains ) {
 
 				// Grab the possible strings for default and any other domain.
 				if ( 'default' === $domain ) {
-					$strings[ $key ][] = __( $value );
-					$strings[ $key ][] = __( strtolower( $value ) );
-					$strings[ $key ][] = __( ucfirst( $value ) );
+					if ( $flags & static::COMPILE_INPUT ) {
+						$strings[ $key ][] = __( $value );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:254]: :no_entry_sign: The $text arg must be a single string literal, not "$value".

<sub>hash: 3ec79062236b134ebbab2ce8bf707925</sub>

File: src/Tribe/I18n.php [hunk 21]
```diff
@@ -190,13 +250,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ( $flags & static::COMPILE_STRTOLOWER ) {
+						$strings[ $key ][] = __( strtolower( $value ) );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:257]: :no_entry_sign: The $text arg must be a single string literal, not "strtolower( $value )".

<sub>hash: 13932b44bb88bdd73ad4c3c3eea08a57</sub>

File: src/Tribe/I18n.php [hunk 22]
```diff
@@ -190,13 +250,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ( $flags & static::COMPILE_UCFIRST ) {
+						$strings[ $key ][] = __( ucfirst( $value ) );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:260]: :no_entry_sign: The $text arg must be a single string literal, not "ucfirst( $value )".

<sub>hash: 3c69bb62cdc1ee2199053ac628e903d5</sub>

File: src/Tribe/I18n.php [hunk 23]
```diff
@@ -190,13 +250,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
 				} else {
-					$strings[ $key ][] = __( $value, $domain );
-					$strings[ $key ][] = __( strtolower( $value ), $domain );
-					$strings[ $key ][] = __( ucfirst( $value ), $domain );
+					if ( $flags & static::COMPILE_INPUT ) {
+						$strings[ $key ][] = __( $value, $domain );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:264]: * :no_entry_sign: The $text arg must be a single string literal, not "$value".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: fa1f463ff83bb98dd799b838c2c8f390</sub>

File: src/Tribe/I18n.php [hunk 24]
```diff
@@ -190,13 +250,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ( $flags & static::COMPILE_STRTOLOWER ) {
+						$strings[ $key ][] = __( strtolower( $value ), $domain );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:267]: * :no_entry_sign: The $text arg must be a single string literal, not "strtolower( $value )".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: dbe4dcaaac22b1b91eb81b5450f86097</sub>

File: src/Tribe/I18n.php [hunk 25]
```diff
@@ -190,13 +250,25 @@ public function compile_translations( array $strings, $domains ) {
...
+					}
+					if ( $flags & static::COMPILE_UCFIRST ) {
+						$strings[ $key ][] = __( ucfirst( $value ), $domain );
```
Comment (tr1b0t) [inline src/Tribe/I18n.php:270]: * :no_entry_sign: The $text arg must be a single string literal, not "ucfirst( $value )".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: bc7a704a3dc2f8bc319a37c62d7a4613</sub>

File: src/Tribe/I18n.php (hunk 20 above)
Comment (tr1b0t) [inline src/Tribe/I18n.php:254]: :no_entry_sign: The $text arg must be a single string literal, not "$value".

<sub>hash: 3ec79062236b134ebbab2ce8bf707925</sub>

File: src/Tribe/I18n.php (hunk 21 above)
Comment (tr1b0t) [inline src/Tribe/I18n.php:257]: :no_entry_sign: The $text arg must be a single string literal, not "strtolower( $value )".

<sub>hash: 13932b44bb88bdd73ad4c3c3eea08a57</sub>

File: src/Tribe/I18n.php (hunk 22 above)
Comment (tr1b0t) [inline src/Tribe/I18n.php:260]: :no_entry_sign: The $text arg must be a single string literal, not "ucfirst( $value )".

<sub>hash: 3c69bb62cdc1ee2199053ac628e903d5</sub>

File: src/Tribe/I18n.php (hunk 23 above)
Comment (tr1b0t) [inline src/Tribe/I18n.php:264]: * :no_entry_sign: The $text arg must be a single string literal, not "$value".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: fa1f463ff83bb98dd799b838c2c8f390</sub>

File: src/Tribe/I18n.php (hunk 24 above)
Comment (tr1b0t) [inline src/Tribe/I18n.php:267]: * :no_entry_sign: The $text arg must be a single string literal, not "strtolower( $value )".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: dbe4dcaaac22b1b91eb81b5450f86097</sub>

File: src/Tribe/I18n.php (hunk 25 above)
Comment (tr1b0t) [inline src/Tribe/I18n.php:270]: * :no_entry_sign: The $text arg must be a single string literal, not "ucfirst( $value )".
* :no_entry_sign: The $domain arg must be a single string literal, not "$domain".

<sub>hash: bc7a704a3dc2f8bc319a37c62d7a4613</sub>

File: src/Tribe/Rewrite.php [hunk 26]
```diff
@@ -706,7 +710,7 @@ protected function get_option_controlled_slug_entry( array $localized_matchers,
 	}
 
 	/**
-	 * Returns the map of localized bases for the specified text domains.
+	 * Rettrns the map of localized bases for the specified text domains.
```
Comment (Camwyn) [inline src/Tribe/Rewrite.php:713]: Returns?
Looks unintentional.

Comment (lucatume) [reply src/Tribe/Rewrite.php:713]: Absolutely not intentional...

**Pull request merged**